### PR TITLE
PR for assigning device to site

### PR DESCRIPTION
**Pull request by abimishr**
**1. Git PR Link:** https://github.com/madhansansel/dnacenter-ansible/pull/250
**2. Problem Description:** Need of assigning devices to site in provisioning module and adding new message for re-provisioning of devices
 **3. Code Changes:**  Added a new parameter called provisioning in the module which when set to false will only assign and when true will do provisioning. Along with that have added the code for different message for re-provisioning.

### CODE REVIEW

File: plugins/modules/provision_workflow_manager.py [hunk 1]
```diff
@@ -268,6 +309,36 @@ def get_dev_type(self):
...
+          - self: The instance of the class containing the 'config' attribute
+                  to be validated.
+        Returns:
+          The method returns an instance of the class with updated attributes:
+          - serial_number: A string indicating the serial number of the device
+        Example:
+          Post creation of the validated input, we this method gets the
+          serial number of the device.
+        """
+
```
[the remark this answers is not in the document]
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py:326]: Removed

File: plugins/modules/provision_workflow_manager.py [hunk 2]
```diff
@@ -268,6 +309,36 @@ def get_dev_type(self):
...
+            op_modifies=True
+        )
+
+        self.log("The device response from 'get_network_device_by_ip' API is {0}".format(str(response)), "DEBUG")
+        dev_dict = response.get("response")
+        serial_number = dev_dict.get("serialNumber")
+
+        self.log("Serial Number of the device is {0}".format(str(serial_number)), "INFO")
+
+        return serial_number
```
[the remark this answers is not in the document]
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py:340]: Added try and except

File: plugins/modules/provision_workflow_manager.py [hunk 3]
```diff
@@ -268,6 +309,36 @@ def get_dev_type(self):
...
+        response = self.dnac_apply['exec'](
+            family="devices",
+            function='get_network_device_by_ip',
+            params={"ip_address": self.validated_config[0]["management_ip_address"]},
+            op_modifies=True
+        )
+
+        self.log("The device response from 'get_network_device_by_ip' API is {0}".format(str(response)), "DEBUG")
+        dev_dict = response.get("response")
+        serial_number = dev_dict.get("serialNumber")
```
[the remark this answers is not in the document]
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py:336]: Added

File: plugins/modules/provision_workflow_manager.py [hunk 4]
```diff
@@ -268,6 +309,36 @@ def get_dev_type(self):
...
+
+        response = self.dnac_apply['exec'](
+            family="devices",
+            function='get_network_device_by_ip',
+            params={"ip_address": self.validated_config[0]["management_ip_address"]},
+            op_modifies=True
+        )
+
+        self.log("The device response from 'get_network_device_by_ip' API is {0}".format(str(response)), "DEBUG")
+        dev_dict = response.get("response")
```
[the remark this answers is not in the document]
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py:335]: Added

File: plugins/modules/provision_workflow_manager.py [hunk 5]
```diff
@@ -268,6 +309,36 @@ def get_dev_type(self):
...
+        Fetches the serial number of the device
+
+        Parameters:
+          - self: The instance of the class containing the 'config' attribute
+                  to be validated.
+        Returns:
+          The method returns an instance of the class with updated attributes:
+          - serial_number: A string indicating the serial number of the device
+        Example:
+          Post creation of the validated input, we this method gets the
```
[the remark this answers is not in the document]
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py:323]: Corrected

File: plugins/modules/provision_workflow_manager.py [hunk 6]
```diff
@@ -268,6 +309,36 @@ def get_dev_type(self):
...
         return device_type
 
+    def get_serial_number(self):
+        """
+        Fetches the serial number of the device
+
+        Parameters:
+          - self: The instance of the class containing the 'config' attribute
+                  to be validated.
+        Returns:
```
[the remark this answers is not in the document]
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py:319]: Added

File: plugins/modules/provision_workflow_manager.py [hunk 7]
```diff
@@ -110,16 +120,46 @@
...
+    dnac_verify: "{{dnac_verify}}"
+    dnac_port: "{{dnac_port}}"
+    dnac_version: "{{dnac_version}}"
+    dnac_debug: "{{dnac_debug}}"
+    dnac_log: True
+    state: merged
+    config:
+        - site_name_hierarchy: Global/USA/San Francisco/BGL_18
+          management_ip_address: 204.192.3.40
+          provisioning: false
```
[the remark this answers is not in the document]
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py:162]: Sure

File: plugins/modules/provision_workflow_manager.py [hunk 8]
```diff
@@ -95,10 +103,12 @@
     post /dna/intent/api/v1/business/sda/provision-device
     post /dna/intent/api/v1/wireless/provision
 
+  - Added 'provisioning' option in v6.13.2
```
[the remark this answers is not in the document]
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py:106]: Corrected

File: plugins/modules/provision_workflow_manager.py [hunk 9]
```diff
@@ -39,9 +39,17 @@
...
         management_ip_address:
-            description: Management Ip Address .
+            description: Management Ip Address of the device.
             type: str
             required: true
+        provisioning:
+            description:
+                - Describes whesther the user wants to perform only site assignment or provisioning of a wired device.
+                - False, for site assignment only
+                - True, for provisioning to a site
```
[the remark this answers is not in the document]
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py:49]: Addressed

File: plugins/modules/provision_workflow_manager.py [hunk 10]
```diff
@@ -39,9 +39,17 @@
...
     suboptions:
         management_ip_address:
-            description: Management Ip Address .
+            description: Management Ip Address of the device.
             type: str
             required: true
+        provisioning:
+            description:
+                - Describes whesther the user wants to perform only site assignment or provisioning of a wired device.
+                - False, for site assignment only
```
[the remark this answers is not in the document]
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py:48]: Addressed

File: plugins/modules/provision_workflow_manager.py [hunk 11]
```diff
@@ -39,9 +39,17 @@
     required: true
     suboptions:
         management_ip_address:
-            description: Management Ip Address .
+            description: Management Ip Address of the device.
             type: str
             required: true
+        provisioning:
+            description:
+                - Describes whesther the user wants to perform only site assignment or provisioning of a wired device.
```
[the remark this answers is not in the document]
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py:47]: Corrected, thanks for pointing it out

File: plugins/modules/provision_workflow_manager.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (madhansansel) [reply plugins/modules/provision_workflow_manager.py:319]: This API does not return any class or update any attributes. It only returns the serial number of devices. Can we simply write as 
`The method returns the serial number of the device as a string. If it fails, it returns None.`

File: plugins/modules/provision_workflow_manager.py [hunk 12]
```diff
@@ -303,14 +387,54 @@ def get_task_status(self, task_id=None):
...
 
+    def get_execution_status(self, execution_id=None):
+        """
+        Fetches the status of the task once any provision API is called
+
+        Parameters:
+          - self: The instance of the class containing the 'config' attribute
+                  to be validated.
+        Returns:
+          The method returns an instance of the class with updated attributes:
```
Comment (madhansansel) [inline plugins/modules/provision_workflow_manager.py:406]: This API does not return any class or update any attributes.

File: plugins/modules/provision_workflow_manager.py [hunk 13]
```diff
@@ -351,6 +475,85 @@ def get_site_type(self, site_name_hierarchy=None):
...
 
+    def get_site_details(self, site_name_hierarchy=None):
+        """
+        Fetches the id and existance of the site
+
+        Parameters:
+          - self: The instance of the class containing the 'config' attribute
+                  to be validated.
+        Returns:
+          The method returns an instance of the class with updated attributes:
```
Comment (madhansansel) [inline plugins/modules/provision_workflow_manager.py:486]: This API does not return any class or update any attributes.

File: plugins/modules/provision_workflow_manager.py [hunk 14]
```diff
@@ -351,6 +475,85 @@ def get_site_type(self, site_name_hierarchy=None):
...
+
+    def get_site_assignment(self):
+        """
+        Fetches the details of devices assigned to a site
+
+        Parameters:
+          - self: The instance of the class containing the 'config' attribute
+                  to be validated.
+        Returns:
+          The method returns an instance of the class with updated attributes:
```
Comment (madhansansel) [inline plugins/modules/provision_workflow_manager.py:528]: This API does not return any class or update any attributes.

**Review by madhansansel — APPROVED** [review]

**Pull request merged**